### PR TITLE
Add inverse_line_graph generator using Roussopolous algorithm.

### PR DESCRIPTION
**Pull request by JamesClough**
This is an initial attempt at implementing an inverse_line_graph operation using the Roussopolous algorithm, as requested in #2139 .
Reviews would be great - I have a few questions about the implementation marked by TODO comments.
Any and all suggestions welcome as the algorithm is reasonably complex and so mistakes on my part are likely and there is probably great potential for making this more efficient.

### CODE REVIEW

File: networkx/generators/inverse_line.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    return P
+    
+def _select_starting_cell(G, starting_edge=None):
+    # If starting edge not specified pick an arbitrary edge - doesn't matter which
+    # However, this function may call itself requiring a specific starting edge
+    if starting_edge == None:
+        # TODO
+        # JC - is this iterator call a good idea? 
+        # JC - If not whats the best way to select an arbitrary edge?
+        e = G.edges().next()
```
Comment (ProgVal) [inline networkx/generators/inverse_line.py]: `e = nx.utils.misc.arbitrary_element(G.edges())`

File: networkx/generators/inverse_line.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+#    Copyright (C) 2016 by
+#    James Clough <james.clough91@gmail.com>
+#    All rights reserved.
+#    BSD license.
+
+__author__ = "\n".join(["James Clough (james.clough91@gmail.com)"])
+
+__all__ = ['inverse_line_graph']
+
+# TODO - is this the right place to put these?
```
Comment (ProgVal) [inline networkx/generators/inverse_line.py]: yes. https://www.python.org/dev/peps/pep-0008/#imports

File: networkx/generators/inverse_line.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    
+    References
+    ----------
+    * Roussopolous, N, "A max {m,n} algorithm for determining the graph H from its line graph G",
+      Information Processing Letters 2, (1973), 108--112. 
+    
+    """
+    
+    # TODO should we have some create_using option here that wraps around the main algorithm?
+    # TODO should we have checks on what kinds of graph are supported or not supported?
```
Comment (ProgVal) [inline networkx/generators/inverse_line.py]: You can use `nx.utils.decorators.not_implemented_for`.

File: networkx/generators/tests/test_inverse_line.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import networkx as nx
+from nose.tools import *
+
+from networkx.testing.utils import *
+# TODO
+# JC: What other tests would we like to see here?
```
Comment (ProgVal) [inline networkx/generators/tests/test_inverse_line.py]: Other types of graphs. Empty graph. Inverse line graph of a graph that is not a line graph (to check an exception of the right type is raised)..

Comment (ProgVal) [inline networkx/generators/tests/test_inverse_line.py]: You can also add test to increase the coverage of your new functions: https://coveralls.io/builds/7739289/source?filename=networkx%2Fgenerators%2Finverse_line.py (red lines are not used by the test suite).

File: networkx/generators/inverse_line.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+            return _select_starting_cell(G, starting_edge=(a,c))
+    else:
+        s = 0
+        odd_triangles = []
+        for T in e_triangles:
+            if _odd_triangle(G, T):
+                s += 1
+                odd_triangles.append(T)
+                
+        if (r==2) and (s==0):
```
Comment (ProgVal) [inline networkx/generators/inverse_line.py]: parenthesis are not needed

File: networkx/generators/inverse_line.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+            # in this case it doesn't matter which of our two triangles we choose, so just use T
+            starting_cell = T 
+        elif (s==r or s==r-1):
```
Comment (ProgVal) [inline networkx/generators/inverse_line.py]: `elif r-1 <= s <= r:` works too

**Comment by JamesClough** [comment]
@ProgVal - I've added more tests - how should I check the test coverage? The link you posted above no longer shows this after the changes. Other than that is there anything else to here apart from tidy up comments?

**Comment by ProgVal** [comment]
Below the the last comment, click “Show all checks”, then click the “Details” link right to coverage/coveralls, and search for your file.

**Comment by JamesClough** [comment]
Thanks - I've added some more test coverage. I'm happy with this now if anyone else wants to review it.

**Comment by jfinkels** [comment]
It will take us some time to review this, since it is somewhat complicated. I have one initial comment from a quick glance at the tests: you are missing the one test that I really care about: testing that `networkx.line_graph` and `networkx.inverse_line_graph` are truly inverses. That is, test for a wide range of graphs, that `line_graph(inverse_line_graph(G)) == inverse_line_graph(line_graph(G)) == G`.

Also, as a rule of thumb for tests, you should only be testing functions that appear in the public interface, so your tests should only call `networkx.inverse_line_graph()`. They should not test the `_triangles()` function, etc. since those are implementation details that (1) are not used by the user, and (2) may change in the future. So if you want to increase test coverage (a good thing to do), you should add tests that only involve the public function `inverse_line_graph()`.

**Comment by JamesClough** [comment]
@jfinkels Thanks - I've removed some unnecessary tests and added in some tests using the existing `networkx.line_graph`. I understand if reviewing takes a while - it is fairly complicated - I'm happy to make changes to notation where that would help here.

**Comment by dschult** [comment]
Merged via #2782

**Pull request closed without merging**